### PR TITLE
Slimmed deep proxy

### PR DESCRIPTION
**Pull request by dai-shi**
From: https://github.com/theKashey/proxyequal/pull/17#issuecomment-489631004

I feel like I'd write a slimmed version of deep proxy.

### CODE REVIEW

**Comment by dai-shi** [comment]
So, the question is whether there's obvious bugs...

File: src/utils.js [hunk 1]
```diff
@@ -5,3 +5,44 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+  };
+  return new Proxy(obj, handler);
+};
+
+const deepChangedCache = new WeakMap();
+export const isDeepChanged = (origObj, nextObj, affected) => {
+  if (origObj === nextObj) return false;
+  if (typeof origObj !== 'object') return true;
+  if (typeof nextObj !== 'object') return true;
+  if (!affected.has(origObj)) return false; // is this safe???
```
Comment (dai-shi) [inline src/utils.js]: This one is the most curious. It behaves differently from `proxyEqual`.

**Comment by dai-shi** [comment]
```
Running benchmark deeptree-nested-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-nested-useReduxState:
┌─────────┬──────────────┬──────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                           │
│         │ (Mount, Avg) │                                                                      │
├─────────┼──────────────┼──────────────────────────────────────────────────────────────────────┤
│ 53.41   │ 145.7, 0.4   │ 60,59,58,59,58,59,57,56,58,57,55,56,50,48,47,48,46,49,45,46,48,45,45 │
└─────────┴──────────────┴──────────────────────────────────────────────────────────────────────┘
Running benchmark deeptree-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-useReduxState:
┌─────────┬──────────────┬──────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                           │
│         │ (Mount, Avg) │                                                                      │
├─────────┼──────────────┼──────────────────────────────────────────────────────────────────────┤
│ 40.21   │ 80.5, 0.6    │ 50,52,53,49,50,49,50,53,54,50,51,50,41,32,31,32,31,32,23,22,24,22,22 │
└─────────┴──────────────┴──────────────────────────────────────────────────────────────────────┘
Running benchmark forms-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark forms-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 57.21   │ 1037.8, 0.3  │ 56,57,56,57,56,59,55,57,58,57,59,58,57,58,57,58,56,57,58,57,58,56,58,58 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark stockticker-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark stockticker-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 58.03   │ 202.9, 0.2   │ 55,59,60,59,60,59,58,59,60,59,60,58,60,59,55,56,57,58,56,58,56,55,54,54 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark tree-view-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark tree-view-useReduxState:
┌─────────┬──────────────┬──────────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                       │
│         │ (Mount, Avg) │                                                                                  │
├─────────┼──────────────┼──────────────────────────────────────────────────────────────────────────────────┤
│ 50.70   │ 537.7, 1.3   │ 40,53,47,49,47,50,52,54,53,47,54,47,50,52,51,52,57,54,55,48,52,43,59,53,43,52,52 │
└─────────┴──────────────┴──────────────────────────────────────────────────────────────────────────────────┘
Running benchmark twitter-lite-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark twitter-lite-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                     │
│         │ (Mount, Avg) │                                                                │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────┤
│ 57.75   │ 2.5, 1.6     │ 59,60,59,60,59,60,59,60,59,60,59,60,59,58,56,59,57,56,50,47,47 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────┘
```

🚀 🚀 🚀 
I hope I'm not misunderstanding anything...

File: src/utils.js [hunk 2]
```diff
@@ -5,3 +5,45 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+  };
+  return new Proxy(obj, handler);
+};
+
+export const isDeepChanged = (origObj, nextObj, affected, cache, depth = 0) => {
+  if (origObj === nextObj) return false;
+  if (typeof origObj !== 'object') return true;
+  if (typeof nextObj !== 'object') return true;
+  if (!affected.has(origObj)) {
+    return depth !== 0; // false for root object, but true for others
```
Comment (dai-shi) [inline src/utils.js]: I'm wondering if this returns false always, we don't need to care about "rest" operator.

I really need help from @theKashey with this stuff...

File: src/utils.js [hunk 3]
```diff
@@ -5,3 +5,45 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+  }
+  if (cache.has(origObj)) {
+    const hit = cache.get(origObj);
+    if (hit.nextObj === nextObj) {
+      return hit.changed;
+    }
+  }
+  const changed = affected.get(origObj)
+    .some(key => isDeepChanged(origObj[key], nextObj[key], affected, cache, depth + 1));
+  cache.set(origObj, { nextObj, changed });
```
Comment (theKashey) [inline src/utils.js]: To break cycles you have to set cache(to undefined) before diving deeper.

File: src/utils.js [hunk 4]
```diff
@@ -5,3 +5,45 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+        affected.set(target, [key]);
+      } else {
+        const used = affected.get(target);
+        if (!used.includes(key)) used.push(key);
+      }
+      const val = target[key];
+      if (typeof val !== 'object') {
+        return val;
+      }
+      if (!proxyMap[key]) {
```
Comment (theKashey) [inline src/utils.js]: proxy map should be external to deepProxy to keep the same req equality between calls, or you will get always unique objects each render.

File: src/utils.js [hunk 5]
```diff
@@ -5,3 +5,45 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+  const handler = {
+    get: (target, key) => {
+      if (!affected.has(target)) {
+        affected.set(target, [key]);
+      } else {
+        const used = affected.get(target);
+        if (!used.includes(key)) used.push(key);
+      }
+      const val = target[key];
+      if (typeof val !== 'object') {
```
Comment (theKashey) [inline src/utils.js]: there is a few objects (like buitins) which count not be proxied.

File: src/utils.js [hunk 6]
```diff
@@ -5,3 +5,45 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+
+export const createDeepProxy = (obj, affected) => {
+  const proxyMap = {};
+  const handler = {
+    get: (target, key) => {
+      if (!affected.has(target)) {
+        affected.set(target, [key]);
+      } else {
+        const used = affected.get(target);
+        if (!used.includes(key)) used.push(key);
```
Comment (theKashey) [inline src/utils.js]: 🚀

File: src/utils.js (hunk 2 above)
Comment (theKashey) [reply src/utils.js]: That might be needed in other cases. What if you `Object.keys(props)`?

File: src/utils.js [hunk 7]
```diff
@@ -5,3 +5,45 @@ export const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayo
...
+    return depth !== 0; // false for root object, but true for others
+  }
+  if (cache.has(origObj)) {
+    const hit = cache.get(origObj);
+    if (hit.nextObj === nextObj) {
+      return hit.changed;
+    }
+  }
+  const changed = affected.get(origObj)
+    .some(key => isDeepChanged(origObj[key], nextObj[key], affected, cache, depth + 1));
```
Comment (theKashey) [inline src/utils.js]: I'll propose use `for` loop here, with early `break` in case of a change.

**Comment by theKashey** [comment]
Looking great, but I urge you to hold it until comprehensive tests written. It took me a while to fix all bugs in `proxyequal` logic.

File: src/utils.js (hunk 2 above)
Comment (dai-shi) [reply src/utils.js]: Oh, it's `Symbol(Symbol.iterator)`. I'm not sure about other cases, but what I could do is to assume the whole object is affected if accessed by a symbol property. (I hesitate writing my own proxy, when I understand that I don't really understand fully...)

File: src/utils.js (hunk 3 above)
Comment (dai-shi) [reply src/utils.js]: Do you mean cycles in state object tree? I need to understand the problem by writing test code...

File: src/utils.js (hunk 4 above)
Comment (dai-shi) [reply src/utils.js]: Totally! I keep forgetting about the ref equality.

File: src/utils.js (hunk 5 above)
Comment (dai-shi) [reply src/utils.js]: I see. I should somehow detect it and work it around. (Or, maybe just warn developers.)

File: src/utils.js (hunk 7 above)
Comment (dai-shi) [reply src/utils.js]: `some` does early break though. Anyway, the for loop is often faster, so I will try when we optimize it (after making sure it works correctly.)

**Comment by dai-shi** [comment]
@theKashey Thanks so much for your comments. I will write some tests. Honestly, it's hard for me to tell if they are comprehensive or not.

File: src/utils.js (hunk 4 above)
Comment (dai-shi) [reply src/utils.js]: And it is not trivial. In the end, I'd reinvent proxyEqual, which I really don't want to do.

File: src/utils.js (hunk 2 above)
Comment (dai-shi) [reply src/utils.js]: And, it's needed for reference equality checking too.

Could we ignore for the root object? Do people do this?
```javascript
  const state = useReduxState();
  return useMemo(() => ({ foo: 'foo' }), [state]);
```
Yeah, maybe?

**Comment by dai-shi** [comment]
@theKashey I did what I can do at this point. If you have time, your comments are so appreciated.

**Comment by dai-shi** [comment]
<details>
<summary>Benchmark with useReduxState 69023f95abde7a7482b1bf77195550358d8f92cb</summary>

```
Running benchmark deeptree-nested-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-nested-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 52.79   │ 124.1, 0.4   │ 58,60,58,59,58,57,55,56,58,54,55,54,47,46,50,46,47,48,50,47,46,45,43,43 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark deeptree-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-useReduxState:
┌─────────┬──────────────┬──────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                           │
│         │ (Mount, Avg) │                                                                      │
├─────────┼──────────────┼──────────────────────────────────────────────────────────────────────┤
│ 40.26   │ 84.9, 0.6    │ 50,52,54,49,52,48,53,51,52,55,51,42,31,32,31,30,31,30,26,22,21,22,22 │
└─────────┴──────────────┴──────────────────────────────────────────────────────────────────────┘
Running benchmark forms-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark forms-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                     │
│         │ (Mount, Avg) │                                                                │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────┤
│ 56.32   │ 1041.9, 0.3  │ 56,55,57,56,57,56,57,56,58,57,58,57,56,55,57,55,56,54,55,56,56 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────┘
Running benchmark stockticker-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark stockticker-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                 │
│         │ (Mount, Avg) │                                                                            │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────────────────┤
│ 56.99   │ 194.1, 0.2   │ 56,59,60,59,60,59,60,59,60,59,58,59,57,59,57,56,57,55,56,55,51,55,53,51,51 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────────────────┘
Running benchmark tree-view-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark tree-view-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                          │
│         │ (Mount, Avg) │                                                                                     │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────────────────┤
│ 50.27   │ 530.5, 1.4   │ 42,50,54,45,48,47,50,51,55,54,47,53,48,46,51,53,57,52,56,48,51,45,47,55,52,45,54,54 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────────────────┘
Running benchmark twitter-lite-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark twitter-lite-useReduxState:
┌─────────┬──────────────┬───────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                    │
│         │ (Mount, Avg) │                                                                               │
├─────────┼──────────────┼───────────────────────────────────────────────────────────────────────────────┤
│ 54.15   │ 3.6, 2.4     │ 59,60,59,60,59,60,59,60,59,60,59,60,58,59,58,57,56,54,55,50,43,44,40,34,33,33 │
└─────────┴──────────────┴───────────────────────────────────────────────────────────────────────────────┘
```
</details>

File: package.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "reactive-react-redux",
   "description": "React Redux binding with React Hooks and Proxy",
-  "version": "2.0.1",
+  "version": "3.0.0-alpha.0",
```
Comment (theKashey) [inline package.json]: 🚀

File: src/utils.js (hunk 6 above)
Comment (dai-shi) [reply src/utils.js]: This turns out to be "really" slow when `used` is big.

**Comment by dai-shi** [comment]
<details>
<summary>Benchmark with useReduxState 8373b9721ef74304d6610d5927a7bb680688c738</summary>

```
Running benchmark deeptree-nested-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-nested-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 53.00   │ 148.8, 0.4   │ 59,58,57,58,59,58,55,56,55,54,57,56,49,47,48,50,48,47,48,45,43,44,45,45 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark deeptree-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                     │
│         │ (Mount, Avg) │                                                                │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────┤
│ 39.07   │ 87.1, 0.6    │ 47,49,50,51,48,47,48,52,50,51,50,43,31,30,31,30,27,22,21,22,22 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────┘
Running benchmark forms-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark forms-useReduxState:
┌─────────┬──────────────┬───────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                            │
│         │ (Mount, Avg) │                                                       │
├─────────┼──────────────┼───────────────────────────────────────────────────────┤
│ 57.14   │ 1049.7, 0.3  │ 55,58,57,58,56,58,56,57,58,60,58,56,57,58,57,58,57,57 │
└─────────┴──────────────┴───────────────────────────────────────────────────────┘
Running benchmark stockticker-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark stockticker-useReduxState:
┌─────────┬──────────────┬───────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                    │
│         │ (Mount, Avg) │                                                                               │
├─────────┼──────────────┼───────────────────────────────────────────────────────────────────────────────┤
│ 53.85   │ 204.1, 0.2   │ 53,59,60,57,58,59,54,58,56,58,52,55,54,55,54,53,48,52,50,51,50,51,47,49,51,51 │
└─────────┴──────────────┴───────────────────────────────────────────────────────────────────────────────┘
Running benchmark tree-view-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark tree-view-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                          │
│         │ (Mount, Avg) │                                                                                     │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────────────────┤
│ 50.05   │ 532.5, 1.3   │ 41,50,57,45,51,44,50,53,55,57,46,53,46,47,52,49,56,52,55,53,46,51,41,55,54,44,45,45 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────────────────┘
Running benchmark twitter-lite-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark twitter-lite-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                 │
│         │ (Mount, Avg) │                                                                            │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────────────────┤
│ 58.82   │ 2.9, 0.8     │ 59,60,59,60,59,60,59,60,59,60,59,60,59,60,59,58,59,60,56,58,55,59,54,59,59 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────────────────┘
```
</details>

File: src/utils.js (hunk 6 above)
Comment (theKashey) [reply src/utils.js]: Have you tried `Set`?

Probably it's time to use some V8-level hacks, like a using different approaches and data structures depending on "size".

Comment (dai-shi) [reply src/utils.js]: Yes, I used `Set`. 8373b9721ef74304d6610d5927a7bb680688c738

**Comment by dai-shi** [comment]
@theKashey https://github.com/dai-shi/reactive-react-redux/issues/1#issuecomment-469877479

If I understand it correctly, it means to use a global `proxyCache` instead of per-hook `proxyCache`s.

```diff
diff --git a/src/useReduxState.js b/src/useReduxState.js
index 8d01436..888173c 100644
--- a/src/useReduxState.js
+++ b/src/useReduxState.js
@@ -69,12 +69,12 @@ export const useReduxStateRich = () => {
   return trapped.state;
 };
 
+const proxyCache = new WeakMap();
 export const useReduxState = (opts = {}) => {
   const forceUpdate = useForceUpdate();
   const store = useContext(ReduxStoreContext);
   const state = store.getState();
   const affected = new WeakMap();
-  const proxyCache = useRef(new WeakMap());
   const lastTracked = useRef(null);
   useIsomorphicLayoutEffect(() => {
     lastTracked.current = {
@@ -109,7 +109,7 @@ export const useReduxState = (opts = {}) => {
     const unsubscribe = store.subscribe(callback);
     return unsubscribe;
   }, [store]); // eslint-disable-line react-hooks/exhaustive-deps
-  return createDeepProxy(state, affected, proxyCache.current);
+  return createDeepProxy(state, affected, proxyCache);
 };
 
 export const useReduxStateSimple = () => {
```

<details>
<summary>Benchmark: baseline with 6a5879bffcbd0e2d655093a1c326ee4759af1577</summary>

```
Running benchmark deeptree-nested-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-nested-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                 │
│         │ (Mount, Avg) │                                                                            │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────────────────┤
│ 52.23   │ 200.1, 0.4   │ 59,57,59,57,58,59,58,56,55,56,55,54,56,51,46,48,47,42,43,49,48,44,46,50,50 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────────────────┘
Running benchmark deeptree-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 34.86   │ 100.6, 0.7   │ 42,45,44,40,31,37,46,37,39,40,36,39,41,42,38,27,25,24,25,27,26,29,30,30 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark forms-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark forms-useReduxState:
┌─────────┬──────────────┬───────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                        │
│         │ (Mount, Avg) │                                                                   │
├─────────┼──────────────┼───────────────────────────────────────────────────────────────────┤
│ 56.76   │ 1095.0, 0.3  │ 55,54,58,55,57,56,55,56,55,56,57,58,57,58,60,58,56,59,57,56,58,58 │
└─────────┴──────────────┴───────────────────────────────────────────────────────────────────┘
Running benchmark stockticker-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark stockticker-useReduxState:
┌─────────┬──────────────┬───────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                    │
│         │ (Mount, Avg) │                                                                               │
├─────────┼──────────────┼───────────────────────────────────────────────────────────────────────────────┤
│ 52.05   │ 240.0, 0.2   │ 53,57,58,56,60,57,54,49,48,50,55,56,52,53,52,54,51,50,49,50,48,50,48,53,47,47 │
└─────────┴──────────────┴───────────────────────────────────────────────────────────────────────────────┘
Running benchmark tree-view-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark tree-view-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                             │
│         │ (Mount, Avg) │                                                                                        │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────────────────────────────┤
│ 50.32   │ 549.2, 1.3   │ 38,51,57,46,48,45,49,55,51,55,43,54,46,52,49,51,50,55,53,54,56,47,50,44,55,53,54,45,45 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────────────────────────────┘
Running benchmark twitter-lite-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark twitter-lite-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 58.90   │ 2.7, 0.8     │ 59,60,59,60,59,60,59,60,59,60,59,60,59,60,59,57,59,56,57,59,57,59,58,58 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
```
</details>

<details>
<summary>Benchmark: with the patch</summary>

```
Running benchmark deeptree-nested-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-nested-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 52.70   │ 147.8, 0.4   │ 58,60,58,59,58,59,57,55,56,57,55,54,48,47,48,46,49,48,49,45,48,42,45,45 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark deeptree-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark deeptree-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 39.29   │ 83.3, 0.6    │ 49,52,45,49,46,49,47,49,51,47,48,50,33,30,31,28,30,31,30,28,22,21,22,22 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark forms-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark forms-useReduxState:
┌─────────┬──────────────┬──────────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                       │
│         │ (Mount, Avg) │                                                                                  │
├─────────┼──────────────┼──────────────────────────────────────────────────────────────────────────────────┤
│ 55.99   │ 1096.6, 0.3  │ 52,49,52,53,55,56,55,57,56,60,57,56,57,56,57,58,57,58,57,55,57,56,58,57,58,57,57 │
└─────────┴──────────────┴──────────────────────────────────────────────────────────────────────────────────┘
Running benchmark stockticker-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark stockticker-useReduxState:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                              │
│         │ (Mount, Avg) │                                                                         │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────┤
│ 56.21   │ 195.2, 0.2   │ 55,59,58,60,58,57,58,57,58,54,57,56,54,56,54,53,55,53,54,55,53,51,52,52 │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────┘
Running benchmark tree-view-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark tree-view-useReduxState:
┌─────────┬──────────────┬────────────────────────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                                                 │
│         │ (Mount, Avg) │                                                                            │
├─────────┼──────────────┼────────────────────────────────────────────────────────────────────────────┤
│ 50.37   │ 529.4, 1.2   │ 40,51,58,45,48,43,49,54,55,46,53,46,50,53,50,56,53,55,49,51,43,51,53,44,44 │
└─────────┴──────────────┴────────────────────────────────────────────────────────────────────────────┘
Running benchmark twitter-lite-useReduxState
    Checking max FPS... (30 seconds)

Results for benchmark twitter-lite-useReduxState:
┌─────────┬──────────────┬──────────────────────────────────────────────────────────┐
│ Avg FPS │ Render       │ FPS Values                                               │
│         │ (Mount, Avg) │                                                          │
├─────────┼──────────────┼──────────────────────────────────────────────────────────┤
│ 59.07   │ 3.6, 0.8     │ 59,60,59,60,59,60,59,60,59,60,59,60,57,59,57,59,58,59,59 │
└─────────┴──────────────┴──────────────────────────────────────────────────────────┘
```
</details>

----

A bit better. (but, not that drastically?)

Another pros is keeping ref equality among hooks, so:
```javascript
const MyComponent = () => {
  const state1 = useReduxState();
  const state2 = useReduxState();
  // state1 === state2
...
```

The cons is the tracking may not be accurate:
```javascript
const ParentComponent = () => {
  const state = useReduxState();
  const foo = state.foo;
  return <ChildComponent foo={foo} />;
};

const ChildComponent = ({ foo }) => {
  const state = useReduxState();
  const foo2 = state.foo;
  return (
    <>
      <Foo foo={foo2} /> 👉 foo2 is tracked in ChildComponent
      <Foo foo={foo} /> 👉  foo is also tracked in ChildComponent although it's from ParentComponent
    </>
  );
};
```

I don't know what would happen if there're async calls.

**Comment by dai-shi** [comment]
```javascript
const Parent = () => {
  const state = useReduxState();
  const foo = state.foo;
  return (
    <>
      <Child1 />
      <Child2 foo={foo} />
    </>
  );
};

const Child1 = () => {
  const state = useReduxState();
  const foo = state.foo;
  return (
    <>
      <Foo foo={foo} /> 👉  foo is tracked in Child1
    </>
  );
};

const Child2 = ({ foo }) => {
  return (
    <>
      <Foo foo={foo} /> 👉  foo is tracked as used in Child1. 😵 
    </>
  );
};
```

I thought it would be problematic in concurrent mode, but even in non-concurrent mode, it may not track as expected.

**Comment by theKashey** [comment]
Yeah. I was thinking about patching `React.createElement` with "deproxify" stuff. That would solve problem of tracking _continuing_ in another component.

**Comment by dai-shi** [comment]
Interesting. Although, with `useReduxState`, we generally want to continue tracking in child components.

**Comment by theKashey** [comment]
> Interesting. Although, with useReduxState, we generally want to continue tracking in child components.

No, this is something you don't want to do
```js
const Parent => ({data}) => (
 <>
   <Child1 data={data} />
   <Child2 data={data} />
 </> 
)

const Child1 = ({data}) => data.stringValue;
const Child2 = ({data}) => data.stringValue[0];
```
💥Parent and Childs would be updated only when __first__ - `[0]` - character is updated. Don't forget - the "usage" is tracking "deepest" access, thus you shall not mix different components, but build a concrete boundaries between them.

**Comment by dai-shi** [comment]
Oh no. How do you deal with it in the first place?
Like in beautiful-react-redux? Ah, it's selector based...

In your example, the string value is not so problematic because we don't track strings.
BUT, I can imagine this:
```javascript
const Parent = () => {
  const state = useReduxState();
  const { foo } = state;
  return (
    <>
      <Child1 foo={foo} />
      <Child2 foo={foo} />
    </>
  );
};

const Child1 = ({ foo }) => (
  <div>{foo.stringValue}</div>
);

const Child2 = ({ foo }) => useMemo(() => (
  <div>{foo.numberValue}</div>
), [foo]);
// Or same thing with React.memo with the second argument
```

I have no idea how to solve this in userland. 🤔 any workaround...

**Comment by dai-shi** [comment]
Let me merge this and I will release it as a new version, hoping someone would try it and give us a feedback.
Your comment is always welcome here or in any other issues/PRs.

**Comment by theKashey** [comment]
Everything is looking good (at least yet 🤓). Look like you nearly fixes a new _patterns_ in React, so many people were struggling with.

**Comment by dai-shi** [comment]
👍 

> a new patterns in React, so many people were struggling with.

Could you elaborate this? what do you mean by a new patterns?

**Comment by theKashey** [comment]
Sorry, 🤯🥴💩. I was trying to point on “patterns” commonly used to wire the app - in your case it’s closer to the magic of Vue (which is also backed by proxies).

There was a long standing issue with redux and selectors, which is just a now issue for you.

So you are fixing old patterns by a new ones. This is something I wanted to say ;)

**Comment by dai-shi** [comment]
Got it!

**Pull request merged**